src/Server.php

```diff
@@ -244,11 +264,88 @@ public function writeError(ConnectionInterface $conn, $code)
             $message .= ': ' . ResponseCodes::$statusTexts[$code];
         }
 
-        $response = new Response($conn);
-        $response->writeHead($code, array(
-            'Content-Length' => strlen($message),
-            'Content-Type' => 'text/plain'
-        ));
-        $response->end($message);
+        $response = new Response(
+            $code,
+            array(
+                'Content-Length' => strlen($message),
+                'Content-Type' => 'text/plain'
+            ),
+            $message
+        );
+
+        $this->handleResponse($conn, $response, '1.1');
+    }
+
+
+    /** @internal */
+    public function handleResponse(ConnectionInterface $connection, ResponseInterface $response, $protocolVersion)
+    {
+        $response = $response->withProtocolVersion($protocolVersion);
+
+        // assign default "X-Powered-By" header as first for history reasons
+        if (!$response->hasHeader('X-Powered-By')) {
+            $response = $response->withHeader('X-Powered-By', 'React/alpha');
+        }
+
+        if ($response->hasHeader('X-Powered-By') && $response->getHeaderLine('X-Powered-By') === ''){
+            $response = $response->withoutHeader('X-Powered-By');
+        }
+
+        $response = $response->withoutHeader('Transfer-Encoding');
+
+        // assign date header if no 'date' is given, use the current time where this code is running
+        if (!$response->hasHeader('Date')) {
+            // IMF-fixdate  = day-name "," SP date1 SP time-of-day SP GMT
+            $response = $response->withHeader('Date', gmdate('D, d M Y H:i:s') . ' GMT');
+        }
+
+        if ($response->hasHeader('Date') && $response->getHeaderLine('Date') === ''){
+            $response = $response->withoutHeader('Date');
+        }
+
+        // assign chunked transfer-encoding if no 'content-length' is given for HTTP/1.1 responses
+        if (!$response->hasHeader('Content-Length') && $protocolVersion === '1.1') {
+            $response = $response->withHeader('Transfer-Encoding', 'chunked');
+        }
+
+        // HTTP/1.1 assumes persistent connection support by default
+        // we do not support persistent connections, so let the client know
+        if ($protocolVersion === '1.1') {
+            $response = $response->withHeader('Connection', 'close');
+        }
+
+        $this->handleResponseBody($response, $connection);
+    }
+
+    private function handleResponseBody(ResponseInterface $response, ConnectionInterface $connection)
+    {
+        $body = $response->getBody();
+        $stream = $body;
+        $bodyString = '';
+        $input = new ReadableStream();
+
+        if (!$response->getBody() instanceof HttpBodyStream) {
+            $stream = $input;
+            $bodyString = $response->getBody()->__toString();
+            $body = new HttpBodyStream($input, null);
+            $response = $response->withBody($body);
+        }
+
+        if ($response->hasHeader('Transfer-Encoding')) {
```

Comment:
Shouldn't be a second check to ensure the contents of this header is `chunked` before we run the body through the chunked encoder? In case we add transfer encoding `X` or `Y` in the future?